File: scenarios/observability.ts
#!/usr/bin/env bun
// scenarios/observability.ts
//
// Proves Track E (observability): from the dev-only Runtime Debug panel an
// operator can, for a connected remote, read live SSH connection diagnostics
// (ping RTT + transport state), per-method RPC metrics (counts/errors/p50/p99),
// one-click a support-diagnostics bundle to the clipboard, and tail the remote
// daemon's log — all without leaving Helmor or shelling into the host.
//
// Each card lives in the same scrollable Settings pane, so the scenario scrolls
// the relevant section to the top before capturing its captioned beat.
//
// Preconditions: `bun run dev` (debug build → bridge + Runtime Debug panel) and
// a connected remote (scripts/setup-remote-workspace.ts). Discovers the remote
// from the registry so it survives renames.

import { Tape } from "./lib.ts";

const NAME = process.env.RUNTIME_NAME ?? "docker-linux-arm64";
const OUT = process.env.TAPE_DIR ?? "./tapes/observability";

const tape = new Tape("observability", OUT);
await tape.connect();

// Confirm a remote is actually connected before we claim to observe it.
const runtimes = (await tape.invoke("list_remote_runtimes", {})) as Array<{
	name: string;
	state?: { type?: string };
}>;
const remote = runtimes.find((r) => r.name === NAME);
tape.assert("remote_connected", remote?.state?.type === "connected", `${NAME}=${remote?.state?.type}`);

/** Scroll the section containing `selector` to the top of the Settings pane. */
async function scrollToSection(selector: string): Promise<boolean> {
	return tape.js<boolean>(
		`var el=document.querySelector(${JSON.stringify(selector)}); if(!el) return false;` +
			`(el.closest('section')||el).scrollIntoView({block:'start',behavior:'auto'}); return true;`,
	);
}

// Clean shell, then open the Runtime Debug panel.
await tape.js('window.location.reload(); return "r";');
await tape.sleep(6000);
await tape.openSettings("runtime-debug");
tape.assert("panel_opens", await tape.waitFor("[role=dialog]"));

// ── Scene 1 — Connection diagnostics (live ping + state) ────────────
tape.assert("diagnostics_card", await tape.waitFor("[data-testid=connection-diagnostics-card]", 10_000));
tape.assert("ping_rtt_shown", await tape.waitFor("[data-testid=diagnostics-ping-ms]", 8000));
await scrollToSection("[data-testid=connection-diagnostics-card]");
await tape.sleep(800);
const diagText = await tape.js<string | null>(
	`var c=document.querySelector('[data-testid=connection-diagnostics-card]'); return c?c.innerText.replace(/\\n+/g," · ").slice(0,200):null;`,
);
tape.log(`diagnostics: ${diagText}`);
await tape.scene({
	caption: `Runtime Debug → Connection diagnostics: live SSH ping RTT, protocol handshake & transport state for ${NAME}`,
	hold: 5,
});

// ── Scene 2 — Per-method RPC metrics table ──────────────────────────
tape.assert("metrics_table", await tape.waitFor("[data-testid=runtime-metrics-table]", 10_000));
await scrollToSection("[data-testid=runtime-metrics-runtime-select]");
await tape.sleep(800);
const methodCount = await tape.js<number>(
	`var t=document.querySelector('[data-testid=runtime-metrics-table] tbody'); return t?t.querySelectorAll('tr').length:0;`,
);
tape.assert("metrics_have_rows", methodCount > 0, `${methodCount} methods`);
await tape.scene({
	caption: "Per-method RPC metrics: call counts, error counts, p50/p99 latency — read straight from the remote daemon",
	hold: 5,
});

// ── Scene 3 — Copy diagnostics bundle (one-click support blob) ──────
await tape.click("[data-testid=runtime-metrics-copy]");
const toast = await tape.waitFor("[data-sonner-toast]", 5000);
tape.assert("copy_toast", toast, toast ? "toast shown" : "no toast (clipboard may be denied in webview)");
await tape.scene({
	caption: "Copy diagnostics: one click bundles health + metrics + the last 50 daemon-log lines into a JSON blob for a support thread",
	record: 2,
	hold: 5,
});

// ── Scene 4 — Daemon log tail ───────────────────────────────────────
tape.assert("daemon_log_pre", await tape.waitFor("[data-testid=daemon-log-pre]", 10_000));
await scrollToSection("[data-testid=daemon-log-runtime-select]");
await tape.sleep(800);
const logLen = await tape.js<number>(
	`var p=document.querySelector('[data-testid=daemon-log-pre]'); return p?p.innerText.split("\\n").length:0;`,
);
tape.assert("daemon_log_has_lines", logLen > 0, `${logLen} lines`);
await tape.scene({
	caption: `Daemon log: tail $HOME/.helmor/server/daemon.log on ${NAME} without SSHing in — the first stop when an agent.send errors`,
	hold: 5,
});

const passed = await tape.finish({ runtimeName: NAME, methodCount, logLen });
process.exit(passed ? 0 : 1);

File: src/scenarios/observability.rs
//! Track E proof: from the dev-only Runtime Debug panel an operator
//! reads live SSH diagnostics (ping RTT + transport state), per-method
//! RPC metrics (counts/errors/p50/p99), one-clicks a support bundle to
//! the clipboard, and tails the remote daemon's log — all without
//! leaving Helmor or shelling into the host.
//!
//! Rust port of `scenarios/observability.ts`. Preconditions: a
//! connected remote runtime named `runtime_name` (defaults to the
//! docker fixture). The scenario discovers the runtime via
//! `list_remote_runtimes` so it survives renames.

use std::time::Duration;

use anyhow::Result;
use serde::Deserialize;
use serde_json::{json, Value};

use crate::tape::{SceneSpec, Tape};

#[derive(Debug, Clone)]
pub struct Config {
    pub runtime_name: String,
}

impl Config {
    pub fn from_env() -> Self {
        Self {
            runtime_name: std::env::var("RUNTIME_NAME")
                .unwrap_or_else(|_| "docker-linux-arm64".into()),
        }
    }
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct RuntimeEntry {
    pub name: String,
    pub state: Option<RuntimeState>,
}

#[derive(Debug, Clone, Default, Deserialize)]
#[serde(default)]
pub struct RuntimeState {
    #[serde(rename = "type")]
    pub kind: Option<String>,
}

pub async fn run(tape: &mut Tape, config: &Config) -> Result<bool> {
    let runtimes: Vec<RuntimeEntry> = tape.invoke("list_remote_runtimes", json!({})).await?;
    let remote = runtimes.iter().find(|r| r.name == config.runtime_name);
    let state_label = remote
        .and_then(|r| r.state.as_ref())
        .and_then(|s| s.kind.as_deref())
        .unwrap_or("<missing>");
    tape.assert(
        "remote_connected",
        state_label == "connected",
        format!("{}={state_label}", config.runtime_name),
    );

    tape.js::<Value>(r#"window.location.reload(); return "r";"#)
        .await?;
    tape.sleep(Duration::from_secs(6)).await;
    tape.open_settings("runtime-debug").await?;
    let panel_opens = tape
        .wait_for("[role=dialog]", Duration::from_secs(5))
        .await?;
    tape.assert("panel_opens", panel_opens, "");

    // Connection diagnostics card.
    let diagnostics_card = tape
        .wait_for(
            "[data-testid=connection-diagnostics-card]",
            Duration::from_secs(10),
        )
        .await?;
    tape.assert("diagnostics_card", diagnostics_card, "");
    let ping_rtt = tape
        .wait_for("[data-testid=diagnostics-ping-ms]", Duration::from_secs(8))
        .await?;
    tape.assert("ping_rtt_shown", ping_rtt, "");
    tape.scroll_to_section("[data-testid=connection-diagnostics-card]")
        .await?;
    tape.sleep(Duration::from_millis(800)).await;
    let diag_text: Option<String> = tape
        .js(
            r#"var c=document.querySelector('[data-testid=connection-diagnostics-card]');
               return c?c.innerText.replace(/\n+/g," · ").slice(0,200):null;"#,
        )
        .await?;
    tape.log(&format!(
        "diagnostics: {}",
        diag_text.clone().unwrap_or_default()
    ));
    tape.scene(
        SceneSpec::new(format!(
            "Runtime Debug → Connection diagnostics: live SSH ping RTT, protocol handshake & transport state for {}",
            config.runtime_name
        ))
        .hold_sec(5),
    )
    .await?;

    // Per-method metrics table.
    let metrics_table = tape
        .wait_for(
            "[data-testid=runtime-metrics-table]",
            Duration::from_secs(10),
        )
        .await?;
    tape.assert("metrics_table", metrics_table, "");
    tape.scroll_to_section("[data-testid=runtime-metrics-runtime-select]")
        .await?;
    tape.sleep(Duration::from_millis(800)).await;
    let method_count: u64 = tape
        .js(
            r#"var t=document.querySelector('[data-testid=runtime-metrics-table] tbody');
               return t?t.querySelectorAll('tr').length:0;"#,
        )
        .await?;
    tape.assert(
        "metrics_have_rows",
        method_count > 0,
        format!("{method_count} methods"),
    );
    tape.scene(
        SceneSpec::new(
            "Per-method RPC metrics: call counts, error counts, p50/p99 latency — read straight from the remote daemon",
        )
        .hold_sec(5),
    )
    .await?;

    // Copy-diagnostics bundle.
    tape.click("[data-testid=runtime-metrics-copy]").await?;
    let toast = tape
        .wait_for("[data-sonner-toast]", Duration::from_secs(5))
        .await?;
    tape.assert(
        "copy_toast",
        toast,
        if toast {
            "toast shown"
        } else {
            "no toast (clipboard may be denied in webview)"
        },
    );
    tape.scene(
        SceneSpec::new(
            "Copy diagnostics: one click bundles health + metrics + the last 50 daemon-log lines into a JSON blob for a support thread",
        )
        .record_sec(2)
        .hold_sec(5),
    )
    .await?;

    // Daemon log tail.
    let daemon_log_pre = tape
        .wait_for("[data-testid=daemon-log-pre]", Duration::from_secs(10))
        .await?;
    tape.assert("daemon_log_pre", daemon_log_pre, "");
    tape.scroll_to_section("[data-testid=daemon-log-runtime-select]")
        .await?;
    tape.sleep(Duration::from_millis(800)).await;
    let log_len: u64 = tape
        .js(
            r#"var p=document.querySelector('[data-testid=daemon-log-pre]');
               return p?p.innerText.split("\n").length:0;"#,
        )
        .await?;
    tape.assert(
        "daemon_log_has_lines",
        log_len > 0,
        format!("{log_len} lines"),
    );
    tape.scene(
        SceneSpec::new(format!(
            "Daemon log: tail $HOME/.helmor/server/daemon.log on {} without SSHing in — the first stop when an agent.send errors",
            config.runtime_name
        ))
        .hold_sec(5),
    )
    .await?;

    tape.finish(json!({
        "runtimeName": config.runtime_name,
        "methodCount": method_count,
        "logLen": log_len,
    }))
    .await
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn runtime_entry_deserializes_connected_state() {
        let v = json!({
            "name": "docker-linux-arm64",
            "state": {"type": "connected"},
        });
        let parsed: RuntimeEntry = serde_json::from_value(v).unwrap();
        assert_eq!(parsed.name, "docker-linux-arm64");
        assert_eq!(
            parsed.state.and_then(|s| s.kind).as_deref(),
            Some("connected")
        );
    }

    #[test]
    fn runtime_entry_with_missing_state_decodes_to_default() {
        let v = json!({"name": "docker-linux-arm64"});
        let parsed: RuntimeEntry = serde_json::from_value(v).unwrap();
        assert!(parsed.state.is_none());
    }

    #[test]
    fn runtime_entry_serde_default_label_is_arm64() {
        unsafe { std::env::remove_var("RUNTIME_NAME") };
        assert_eq!(Config::from_env().runtime_name, "docker-linux-arm64");
    }
}
